### PR TITLE
ci: run USDT interface tests in the CI

### LINKED ISSUE #23296: tracing: tests for USDT tracepoints
#22006 added the first three USDT based tracepoints to Bitcoin Core. To provide a [semi-stable](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#semi-stable-api) tracepoint API the tracepoints need test coverage. The tracepoints can be tested in the functional tests using the Python wrapper of [BCC](https://github.com/iovisor/bcc). 

Before adding more tracepoints, the existing three tracepoints from #22006 should be tested.

Notes:
1. We currently only support the tracepoints on Linux. The tests should be skipped on other operating systems.
2. Hooking into the tracepoints via the Linux kernel requires special privileges. Since kernel version 5.8. (Aug. 2020) the `CAP_BPF` can be used. On older kernel version the overloaded catch-all capability `CAP_SYS_ADMIN` is required. Functional tests shouldn't require `CAP_SYS_ADMIN` as that essentially means running the test suite with `root` privileges.  
3. The tests require the BCC Python library. This should be an optional dependency. Tests should be skipped if the dependency isn't present. 

The `connect_block` tracepoint can be tested by mining blocks with transactions and checking that the tracepoint passes the correct data. The `net` `inbound_message` and `outbound_message` tracepoints can be tested by checking the traffic between two nodes.

### LINKED ISSUE #24782: CI, tracing: run tracepoint interface tests in the CI
The first tracepoint tests were merged in #24358. However, they are skipped in the CI. I've already done some experiments running them in the CI in https://github.com/bitcoin/bitcoin/issues/23296, however I haven't been successful with it yet. The main issue was the limited capabilities inside the CirrusCI docker container.

An option would be to ask CirrusCI for help, another would be to run this part in a VM, if it's not to much overhead.

### PR DESCRIPTION
**Pull request by 0xB10C**
Changes a CI task that runs test the previously not run `test/functional/interface_usdt_*.py` functional tests (added in https://github.com/bitcoin/bitcoin/pull/24358). 

This task is run as CirussCI `compute_engine_instance` VM as hooking into the tracepoints is not possible in CirrusCI docker containers (https://github.com/bitcoin/bitcoin/issues/23296#issuecomment-1024920845). We use an unoffical PPA and untrusted  `bpfcc-tools` package in the CI as the Ubuntu jammy and Debian bullseye packages are outdated. We hope use an official package when new Ubuntu/Debian releases are available for the use with Google Compute Engine.

We make sure to hook into `bitcoind` binaries in USDT interface tests via their PID, instead of their path. This makes sure multiple functional tests running in parallel don't interfere with each other.

The utxocache USDT interface tests is adopted to a change of the functional test framework that wasn't detected as the tests weren't run in the CI. As the tracepoints expose internals, it can happen that we need to adopt the interface test when internals change. This is a bit awkward, and if it happens to frequently, we should consider generalizing the tests a bit more. For now it's fine, I think.

See the individual commit messages for more details on the changes.

Fixes https://github.com/bitcoin/bitcoin/issues/24782
Fixes https://github.com/bitcoin/bitcoin/issues/23296

I'd like to hear from reviewers:
- Are we OK with using the [`hadret/bpfcc`](https://launchpad.net/~hadret/+archive/ubuntu/bpfcc) PPA for now? There is a clear plan when to drop it and as is currently, it could only impact the newly added VM task. 
- ~~Adding a new task increases CI runtime and costs. Should an existing `container` CI task be ported to a VM and reused instead?~~ Yes, see https://github.com/bitcoin/bitcoin/pull/25528#issuecomment-1179509525

### CODE REVIEW

**Comment by 0xB10C** [comment]
> With the only change being to switch to debian:bookworm, see [MarcoFalke/bitcoin-core-with-ci@803f365#diff-b37d362747ecdb36dbfcbe2b8aab779f06404eee471ad835b58355e353d4fe79R9](https://github.com/MarcoFalke/bitcoin-core-with-ci/commit/803f365f47d2f40a88d8969f0e84a6779dc5a5cb#diff-b37d362747ecdb36dbfcbe2b8aab779f06404eee471ad835b58355e353d4fe79R9)
> 
> Any ideas?

Interesting. This seems to be unrelated to the tests. More the tracepoint macros in general. Probably a problem with the Systemtap package (`systemtap-sdt-dev`) in bookworm? Per https://packages.debian.org/bookworm/systemtap-sdt-dev it's version 4.7, which is also what we use in depends since https://github.com/bitcoin/bitcoin/pull/25360. Does a depends build on bookworm fail too?

I'll also do some poking.

**Comment by 0xB10C** [comment]
Tried running a depends build on bookworm - but no success so far: https://cirrus-ci.com/task/5255417889554432?logs=ci#L570

https://github.com/0xB10C/bitcoin/commit/5641883b3a148f01e1ef910705fd82a4b5132dde

**Comment by MarcoFalke** [comment]
In any case it would fail because it wouldn't be possible to install the kernel headers in docker: https://cirrus-ci.com/task/5345925198512128?logs=ci#L3567 ?

**Comment by MarcoFalke** [comment]
Makes sense. That doesn't seem to be worth the effort.

An alternative to using a private ppa would be to bump the OS version in `/etc/apt/sources.list` before installing the package (See for example https://github.com/google/oss-fuzz/commit/0da6949a7bc3586cbf160738e8f09b209ea97a06 ), though that'd just seem too brittle.

To preserve CI resources, I am thinking about combining this with another task. Since a wallet would be good to also cover the coinselection traces, and a recent-ish OS is required, we could re-use one of the sanitizer builds or the i686 build?

**Comment by 0xB10C** [comment]
As discussed offline with MarcoFake, we're using the `[ASan + LSan + UBSan + integer, no depends, USDT] [jammy]` task to run the USDT tests. The i686 didn't work - probably because the tracing isn't really supported on 32bit.

**Comment by MarcoFalke** [comment]
cr ACK cc7335edc87c6ef34429b4df94f53973db520aac

File: .cirrus.yml [hunk 1]
```diff
@@ -259,12 +259,20 @@ task:
...
   << : *GLOBAL_TASK_TEMPLATE
-  container:
-    image: ubuntu:jammy
+  # We can't use a 'container' for the USDT interface tests as the CirrusCI
+  # containers don't have privileges to hook into bitcoind. CirrusCI uses
+  # Google Compute Engine instances: https://cirrus-ci.org/guide/custom-vms/
+  # Images can be found here: https://cloud.google.com/compute/docs/images/os-details
+  compute_engine_instance:
+    image_project: ubuntu-os-cloud
+    image: family/ubuntu-2204-lts # when upgrading, check if we can drop "ADD_UNTRUSTED_BPFCC_PPA"
```
Comment (MarcoFalke) [inline .cirrus.yml:270]: note to the future: I was able to change this to `family/ubuntu-2110`, but the gce docs only include lts version. So if `2210` has the updated package, we can probably switch to that: https://packages.ubuntu.com/kinetic/bpfcc-tools

**Comment by theStack** [comment]
Concept ACK

File: ci/test/04_install.sh [hunk 2]
```diff
@@ -68,6 +68,14 @@ if [[ $DOCKER_NAME_TAG == *centos* ]]; then
   ${CI_RETRY_EXE} CI_EXEC dnf -y install epel-release
   ${CI_RETRY_EXE} CI_EXEC dnf -y --allowerasing install "$DOCKER_PACKAGES" "$PACKAGES"
 elif [ "$CI_USE_APT_INSTALL" != "no" ]; then
+  if [[ "${ADD_UNTRUSTED_BPFCC_PPA}" == "true" ]]; then
+    # Ubuntu 22.04 LTS and Debian 11 both have an outdated bpfcc-tools packages.
+    # The iovisor PPA is outdated as well. The next Ubuntu and Debian releases will contain updated
+    # packages. Meanwhile, use an untrusted PPA to install an up-to-date version of the bpfcc-tools
+    # package.
+    # TODO: drop this once we can use newer images in GCE
+    CI_EXEC add-apt-repository ppa:hadret/bpfcc
```
Comment (hebasto) [inline ci/test/04_install.sh:77]: Using of `add-apt-repository` [requires](https://github.com/bitcoin/bitcoin/pull/26571) some packages been installed.

File: ci/test/00_setup_env_native_tidy.sh [hunk 3]
```diff
@@ -15,5 +15,5 @@ export RUN_FUNCTIONAL_TESTS=false
 export RUN_FUZZ_TESTS=false
 export RUN_TIDY=true
 export GOAL="install"
-export BITCOIN_CONFIG="CC=clang CXX=clang++ --with-incompatible-bdb --disable-hardening CFLAGS='-O0 -g0' CXXFLAGS='-O0 -g0'"
+export BITCOIN_CONFIG="CC=clang CXX=clang++ --enable-c++20 --with-incompatible-bdb --disable-hardening CFLAGS='-O0 -g0' CXXFLAGS='-O0 -g0'"
```
Comment (hebasto) [inline ci/test/00_setup_env_native_tidy.sh:18]: @0xB10C @MarcoFalke 

What was the point to add `--enable-c++20` here?

Asking in context of https://github.com/bitcoin/bitcoin/pull/26763#discussion_r1058860943.

Comment (MarcoFalke) [reply ci/test/00_setup_env_native_tidy.sh:18]: It wasn't added in this pull, see the diff in `ci/test/00_setup_env_native_asan.sh`.

It was added to check that compiling with it works.

Comment (hebasto) [reply ci/test/00_setup_env_native_tidy.sh:18]: Could another CI task use it instead of the "tidy" one? (fwiw, we have a native Windows build using C++20)

Comment (MarcoFalke) [reply ci/test/00_setup_env_native_tidy.sh:18]: Sure, any Linux task that builds all targets (including fuzz and gui) should do. For example, multiprocess, asan, or tsan.

Comment (hebasto) [reply ci/test/00_setup_env_native_tidy.sh:18]: > It wasn't added in this pull, see the diff in `ci/test/00_setup_env_native_asan.sh`.

Suggesting to revert the move of `--enable-c++20` between CI tasks in #26770.

**Pull request merged**